### PR TITLE
Session aware logout, backend logout url approach

### PR DESCRIPTION
**Pull request by babs**
## Description

This PR implements backend logout approach, might be better than #1875 and might also satisfy #884

## Motivation and Context

Implement backend logout.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

**Comment by babs** [comment]
I can't do much for `legacyProviderFlagSet` LoC count :/

**Comment by babs** [comment]
Still relevant;)

**Comment by JoelSpeed** [comment]
Apologies for not getting to this, I hope to review soon, looks like there are a couple of conflicts in the keycloak code at the moment, do we want to get those resolved?

**Comment by babs** [comment]
Conflicts resolved, ready to merge.

**Comment by babs** [comment]
Not stale

**Comment by gchait** [comment]
Thank you for that, would love to see it merged soon!

**Comment by babs** [comment]
Still relevent

**Comment by babs** [comment]
.

**Comment by babs** [comment]
.

**Comment by babs** [comment]
.

File: docs/docs/configuration/overview.md [hunk 1]
```diff
@@ -81,6 +81,7 @@ An example [oauth2-proxy.cfg](https://github.com/oauth2-proxy/oauth2-proxy/blob/
 | `--auth-logging-format` | string | Template for authentication log lines | see [Logging Configuration](#logging-configuration) |
 | `--authenticated-emails-file` | string | authenticate against emails via file (one per line) | |
 | `--azure-tenant` | string | go to a tenant-specific or common (tenant-independent) endpoint. | `"common"` |
+| `--backend-logout-url` | string | URL to call to perform backend logout, `{id_token}` would be replaced by the actual `id_token` if available in the session | |
```
Comment (kvanzuijlen) [inline docs/docs/configuration/overview.md:84]:
```suggestion
| `--backend-logout-url` | string | URL to call to perform backend logout, `{id_token}` in the url would be replaced by the actual `id_token` if available in the session | |
```

File: oauthproxy.go [hunk 2]
```diff
@@ -725,6 +725,24 @@ func (p *OAuthProxy) SignOut(rw http.ResponseWriter, req *http.Request) {
...
+			if err != nil {
+				logger.Errorf("error while calling backend logout: %v", err)
+			} else {
+				defer resp.Body.Close()
+				if resp.StatusCode != 200 {
+					logger.Errorf("error while calling backend logout, %v returned a %v", backendLogoutURL, resp.StatusCode)
+				}
+			}
+		}
+	}
```
Comment (kvanzuijlen) [inline oauthproxy.go:744]: I think this can be simplified if you refactored this to be in a function. That way you'd be able to use the "return early" pattern. Right now this adds quite a lot of complexity due too the many if-statements.

**Review by kvanzuijlen — COMMENTED** [review]
@babs Thanks for the PR!

Regarding the code-climate issue, that might be fixed if you merge with the latest master. @tuunit @JoelSpeed anything I missed?

File: oauthproxy.go (hunk 2 above)
Comment (tuunit) [reply oauthproxy.go:744]: Please reduce the `if` nesting and isolate it into its own function as suggest by @kvanzuijlen.

Something like:

```golang
p.BackendLogout(rw, req)

func (p *OAuthProxy) BackendLogout(rw http.ResponseWriter, req *http.Request) {
	session, err := p.getAuthenticatedSession(rw, req)	

	if err != nil || session == nil {
		// error handling is missing here
		// at least logging
		return;
	}

	providerData := p.provider.Data()
	if providerData.BackendLogoutURL == "" {
		return;
	}

	backendLogoutURL := strings.ReplaceAll(providerData.BackendLogoutURL, "{id_token}", session.IDToken)
	resp, err := http.Get(backendLogoutURL) // #nosec G107

	if err != nil {
		logger.Errorf("error while calling backend logout: %v", err)
	} else {
		defer resp.Body.Close()

		if resp.StatusCode != 200 {
			logger.Errorf("error while calling backend logout, %v returned a %v", backendLogoutURL, resp.StatusCode)
		}
	}
}
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -8,11 +8,12 @@
 
 ## Changes since v7.4.0
 
+- [#1839](https://github.com/oauth2-proxy/oauth2-proxy/pull/1839) Add readiness checks for deeper health checks (@kobim)
```
Comment (tuunit) [inline CHANGELOG.md:11]: why did you move that line? please revert :)

File: providers/keycloak_oidc.go [hunk 4]
```diff
@@ -21,6 +21,10 @@ func NewKeycloakOIDCProvider(p *ProviderData, opts options.KeycloakOptions) *Key
 		name: keycloakOIDCProviderName,
 	})
 
+	if opts.BackendLogoutURL != "" {
+		p.BackendLogoutURL = opts.BackendLogoutURL
+	}
```
Comment (tuunit) [inline providers/keycloak_oidc.go:26]: I don't think the check is necessary.
```suggestion
	p.BackendLogoutURL = opts.BackendLogoutURL
```

**Review by tuunit — CHANGES_REQUESTED** [review]

File: pkg/apis/options/legacy_options.go [hunk 5]
```diff
@@ -588,6 +589,7 @@ func legacyProviderFlagSet() *pflag.FlagSet {
 	flagSet.String("user-id-claim", OIDCEmailClaim, "(DEPRECATED for `oidc-email-claim`) which claim contains the user ID")
 	flagSet.StringSlice("allowed-group", []string{}, "restrict logins to members of this group (may be given multiple times)")
 	flagSet.StringSlice("allowed-role", []string{}, "(keycloak-oidc) restrict logins to members of these roles (may be given multiple times)")
+	flagSet.String("backend-logout-url", "", "url to call to perfrom a backend logout, {id_token} can be used as placeholder for the id_token")
```
Comment (tuunit) [inline pkg/apis/options/legacy_options.go:592]:
```suggestion
	flagSet.String("backend-logout-url", "", "url to perform a backend logout, {id_token} can be used as placeholder for the id_token")
```

File: docs/docs/configuration/overview.md (hunk 1 above)
Comment (tuunit) [inline docs/docs/configuration/overview.md:84]:
```suggestion
| `--backend-logout-url` | string | URL to perform backend logout, if you use `{id_token}` in the url it will be replaced by the actual `id_token` of the user session | |
```

**Review by tuunit — CHANGES_REQUESTED** [review]

File: docs/docs/configuration/overview.md (hunk 1 above)
Comment (babs) [reply docs/docs/configuration/overview.md]: rephrase suggested by tuunit includes this change

File: oauthproxy.go (hunk 2 above)
Comment (babs) [reply oauthproxy.go]: implemented :)

**Review by babs — COMMENTED** [review]
AFAIK, all suggestions have been implemented

File: providers/keycloak_oidc.go [hunk 6]
```diff
@@ -21,6 +21,8 @@ func NewKeycloakOIDCProvider(p *ProviderData, opts options.Provider) *KeycloakOI
 		name: keycloakOIDCProviderName,
 	})
 
+	p.BackendLogoutURL = opts.KeycloakConfig.BackendLogoutURL
+
```
Comment (tuunit) [inline providers/keycloak_oidc.go:25]: Final suggestion, move this to `setProviderDefaults(defaults providerDefaults)` in `provider_data.go`

File: oauthproxy.go [hunk 7]
```diff
@@ -742,9 +742,37 @@ func (p *OAuthProxy) SignOut(rw http.ResponseWriter, req *http.Request) {
...
+	backendLogoutURL := strings.ReplaceAll(providerDatas.BackendLogoutURL, "{id_token}", session.IDToken)
+	resp, err := http.Get(backendLogoutURL) // #nosec G107
+	if err != nil {
+		logger.Errorf("error while calling backend logout: %v", err)
+	} else {
+		defer resp.Body.Close()
+		if resp.StatusCode != 200 {
+			logger.Errorf("error while calling backend logout, %v returned a %v", backendLogoutURL, resp.StatusCode)
+		}
+	}
```
Comment (tuunit) [inline oauthproxy.go:773]:
```suggestion
	session, err := p.getAuthenticatedSession(rw, req)
	if err != nil {
		logger.Errorf("error getting authenticated session during backend logout: %v", err)
		return
	}
	
	if session == nil {
		return
	}
	
	providerData := p.provider.Data()
	if providerData.BackendLogoutURL == "" {
		return
	}
	
	backendLogoutURL := strings.ReplaceAll(providerData.BackendLogoutURL, "{id_token}", session.IDToken)
	resp, err := http.Get(backendLogoutURL) // #nosec G107
	if err != nil {
		logger.Errorf("error while calling backend logout: %v", err)
		return
	}
	
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		logger.Errorf("error while calling backend logout url '%v': returned error code %v", backendLogoutURL, resp.StatusCode)
	}
```

File: oauthproxy.go [hunk 8]
```diff
@@ -742,9 +742,37 @@ func (p *OAuthProxy) SignOut(rw http.ResponseWriter, req *http.Request) {
...
+	if session == nil {
+		return
+	}
+	providerDatas := p.provider.Data()
+	if providerDatas.BackendLogoutURL == "" {
+		return
+	}
+	backendLogoutURL := strings.ReplaceAll(providerDatas.BackendLogoutURL, "{id_token}", session.IDToken)
+	resp, err := http.Get(backendLogoutURL) // #nosec G107
```
Comment (tuunit) [inline oauthproxy.go:765]: is the `nosec G107` really necessary? can we somehow avoid this?

File: pkg/apis/options/providers.go [hunk 9]
```diff
@@ -138,6 +138,9 @@ type KeycloakOptions struct {
 
 	// Role enables to restrict login to users with role (only available when using the keycloak-oidc provider)
 	Roles []string `json:"roles,omitempty"`
+
+	// URL to call to perform backend logout, `{id_token}` would be replaced by the actual `id_token` if available in the session
+	BackendLogoutURL string `json:"backendLogoutURL"`
```
Comment (tuunit) [inline pkg/apis/options/providers.go:143]: remove from keycloak options only use the generic one in provider struct

**Review by tuunit — CHANGES_REQUESTED** [review]

File: oauthproxy.go (hunk 8 above)
Comment (babs) [reply oauthproxy.go:765]: I don't think it's doable, in anycase, it would be a variable but not a end user input  that can be tampered with (more an admin input)

Comment (tuunit) [reply oauthproxy.go:765]: Okay leave it as is for now :)

File: oauthproxy.go (hunk 7 above)
Comment (babs) [reply oauthproxy.go]: implemented

File: providers/keycloak_oidc.go (hunk 6 above)
Comment (babs) [reply providers/keycloak_oidc.go]: done

File: pkg/apis/options/providers.go (hunk 9 above)
Comment (babs) [reply pkg/apis/options/providers.go]: done

**Review by tuunit — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 10]
```diff
@@ -43,6 +43,7 @@ will remove this option. This change makes sure that the configuration options f
 N/A
 
 ## Changes since v7.5.0
+- [#1876](https://github.com/oauth2-proxy/oauth2-proxy/pull/1876) Add `--backend-logout-url` with `{id_token}` placeholder (@babs)
```
[the remark this answers is not in the document]
Comment (babs) [reply CHANGELOG.md]: done :)

**Review by tuunit — APPROVED** [review]

**Comment by andrewkcarter** [comment]
Money. I was just looking for a solution to this.

File: oauthproxy.go (hunk 8 above)
Comment (JoelSpeed) [reply oauthproxy.go:768]: I would appreciate a code comment explaining why we are having to use the gosec so that we can remember this context in the future

File: oauthproxy.go [hunk 11]
```diff
@@ -742,9 +742,41 @@ func (p *OAuthProxy) SignOut(rw http.ResponseWriter, req *http.Request) {
...
+	backendLogoutURL := strings.ReplaceAll(providerData.BackendLogoutURL, "{id_token}", session.IDToken)
+	resp, err := http.Get(backendLogoutURL) // #nosec G107
+	if err != nil {
+		logger.Errorf("error while calling backend logout: %v", err)
+		return
+	}
+
+	defer resp.Body.Close()
+	if resp.StatusCode != 200 {
+		logger.Errorf("error while calling backend logout url '%v': returned error code %v", backendLogoutURL, resp.StatusCode)
```
Comment (JoelSpeed) [inline oauthproxy.go:776]: Nit, should use `%q` for a quoted string formatting
```suggestion
		logger.Errorf("error while calling backend logout url %q: returned error code %v", backendLogoutURL, resp.StatusCode)
```

**Review by JoelSpeed — COMMENTED** [review]
Is it possible, for any of the providers, to set a default for the backend logout URL, or are we assuming that setting a default would be a bad default and that most wouldn't actually want that?

**Comment by babs** [comment]
There is no standard for background logout, so I'd say option 2 :)

**Comment by babs** [comment]
Rebased against master

**Comment by babs** [comment]
Rebased against master

**Review by JoelSpeed — APPROVED** [review]

**Comment by andrewkcarter** [comment]
Excellent. When is the next release scheduled?

**Comment by redboxstudio** [comment]
Hello, please write some documentation, the description is scant, what should I do about it?  do I understand correctly that the application should follow http://app.app.example/logout where nginx location follows in 
 this link http://localhost:4180/oauth2/sign_out?rd=http://localhost:8080/auth/realms/testrealm/protocol/openid-connect/logout?id_token_hint={id_token}%26post_logout_redirect_uri=http:// localhost:4180/
what then should be in the backend-logout-url

**Comment by babs** [comment]
Several logout methods has been considered but as `id_token` is a pretty sensitive information, backend logout has beeen privilegied.

You have to consider 2 course of actions when you send the browser to the `sign_out` endpoint, running in parallel:
- Client side you want the browser to be redirected on a gien url (the rd parameter in qs)
- Server to server call from oauth2-proxy to your oidc provider (keycloak) through a secure channel with pattern replacement containing the sensitive info

So http://localhost:8080/auth/realms/testrealm/protocol/openid-connect/logout?id_token_hint={id_token} goes into `--backend-logout-url` and http://localhost:4180/ goes to `rd` argument of sign_out.

**Comment by ch08532** [comment]
> Hello, please write some documentation, the description is scant, what should I do about it? do I understand correctly that the application should follow http://app.app.example/logout where nginx location follows in this link [http://localhost:4180/oauth2/sign_out?rd=http://localhost:8080/auth/realms/testrealm/protocol/openid-connect/logout?id_token_hint={id_token}%26post_logout_redirect_uri=http://](http://localhost:4180/oauth2/sign_out?rd=http://localhost:8080/auth/realms/testrealm/protocol/openid-connect/logout?id_token_hint=%7Bid_token%7D%26post_logout_redirect_uri=http://) localhost:4180/ what then should be in the backend-logout-url

I was able to get this to work with NGINX proxy and Keycloak OIDC provider:


OAuth2proxy config file entry:

```
backend_logout_url="http://keycloak.local/realms/redeye/protocol/openid-connect/logout?post_logout_redirect_uri=https%3A%2F%2Fapp.local&id_token_hint={id_token}"
```

NGINX entry:

```
# Logout route
location /logout {
        proxy_pass http://oauth2-proxy:4180/oauth2/sign_out; 

        include       /etc/nginx/proxy_params;
        proxy_set_header Cookie                  $http_cookie;  # Forward session cookies
    }
```

**Comment by babs** [comment]
Good news.
But again you don't need to specify `post_logout_redirext_uri` in oauth2-proxy's `backend_logout_url` option.

**Comment by ch08532** [comment]
> Good news. But again you don't need to specify `post_logout_redirect_uri` keycloak parameter in oauth2-proxy's `backend_logout_url` option.

Ok thank you for the response.  I tested this and it works as you described.  

```
backend_logout_url="http://keycloak.local/realms/redeye/protocol/openid-connect/logout?id_token_hint={id_token}"`
```

**Comment by babs** [comment]
Glad to hear that ;)

**Comment by redboxstudio** [comment]
> Хорошие новости. Но опять же, вам не нужно указывать `post_logout_redirect_uri`параметр keycloak в опции oauth2-proxy `backend_logout_url`.
Why?  How to back in  app?
mb  this need add to  nginx conf  ? 
location /logout {
        proxy_pass http://oauth2-proxy:4180/oauth2/sign_out?rd=https://myapp.local;

**Comment by redboxstudio** [comment]
> Хорошие новости. Но опять же, вам не нужно указывать `post_logout_redirect_uri`параметр keycloak в опции oauth2-proxy `backend_logout_url`.

Why?  How to back in  app?
mb  this need add to  nginx conf  ? 
location /logout {
        proxy_pass http://oauth2-proxy:4180/oauth2/sign_out?rd=https://myapp.local;

**Comment by redboxstudio** [comment]
> Good news. But again you don't need to specify `post_logout_redirect_uri` keycloak parameter in oauth2-proxy's `backend_logout_url` option.

Why?  How to back in  app?
mb  this need add to  nginx conf  ? 
location /logout {
        proxy_pass http://oauth2-proxy:4180/oauth2/sign_out?rd=https://myapp.local;

**Comment by babs** [comment]
More like that, yes ;)
